compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -51,41 +51,108 @@ static bool hasIntrinsic(IREE::GPU::TargetAttr target,
+  // tends to be larger than the A and B matrix tiles.
+  const int kMaxAccumulatorRegisterBits = 256 * 32;
+
+  MMAAttr intrinsicMma = MMAAttr::get(ctx, *intrinsic);
+  auto [intrinsicA, intrinsicB, intrinsicC] = intrinsicMma.getABCVectorTypes();
+  // The unrollK factor serves to allow loads from the A and B matrices to use
+  // the target ISA's vector loads. For instance, if the ISA has 128-bit loads
+  // and each intrinsic consumes only 32 bits from A and B, then we want to set
+  // unrollK=4 to turn 4 separate 32-bit loads into one 128-bit load.
+  const int unrollK =
+      kLoadInstructionBits /
+      std::min(
+          intrinsicA.getElementTypeBitWidth() * intrinsicA.getNumElements(),
+          intrinsicB.getElementTypeBitWidth() * intrinsicB.getNumElements());
+  // The total amount of unrolling along the M and N dimensions is normally
+  // limited only by the number of available registers, since larger M and N
+  // yields higher arithmetic intensity. Here, we do not yet distinguish between
+  // plain unrolling (more instructions on each thread) and
+  // unrolling-to-subgroups (more threads).
+  const int totalUnrollMN =
+      kMaxAccumulatorRegisterBits /
+      (intrinsicC.getElementTypeBitWidth() * intrinsicC.getNumElements());
+  const int totalUnrollM = static_cast<int>(
+      std::floor(std::sqrt(static_cast<float>(totalUnrollMN))));
+  const int totalUnrollN = totalUnrollMN / totalUnrollM;
+  // Now we introduce unroll-to-subgroups. It doesn't change the overall tile
+  // size, as it increases the number of subgroups but correspondingly decreases
+  // the number of registers available to each subgroups. In other words, the
+  // overall tile size determined above only needed to be concerned with the
+  // overall number of registers, not with how they are split between subgroups.
+  //
+  // For now for simplicity we put all the unroll-to-subgroups in the N
+  // dimension. That might be suboptimal, revisit later. That does simplify the
+  // below adjustments for narrow M/N, as we don't need to think about
+  // unroll-to-subgroups when making the narrowing adjustment.
+  int unrollMToSubgroups = 1;
+  int unrollNToSubgroups = kMaxSubgroups;
+  int unrollM = totalUnrollM / unrollMToSubgroups;
+  int unrollN = totalUnrollN / unrollNToSubgroups;
+
+  //
+  // Step 3: Adjust the unrolling factors when there is a narrow dimension.
+  //
+  IREE::Encoding::MatmulNarrowDim narrowDim =
+      IREE::Encoding::getMatmulNarrowDim(encoding);
+  if (narrowDim.isM()) {
+    unrollM = std::min<int>(
+        unrollM, llvm::divideCeil(narrowDim.size, intrinsicMma.getMSize()));
   }
-  if (auto m = match(MMAIntrinsic::MFMA_I32_16x16x32_I8, 8, 1, 2, 4, 2)) {
-    return m;
+  if (narrowDim.isN()) {
+    std::swap(unrollM, unrollN);
+    std::swap(unrollMToSubgroups, unrollNToSubgroups);
+    assert(unrollNToSubgroups == 1);
+    unrollN = std::min<int>(
```

Comment:
You shouldn't specify the template parameter type in these functions. Instead, cast the operands.